File: web-ui/src/api/session-views.ts
import { apiFetch } from "./client";
import type { OpenCodeEvent } from "../types";
import type { SessionStats } from "./state";

// ── Session events SSE ────────────────────────────────

export function createSessionEventsSSE(): EventSource {
  // Cookie auth: browser sends opman_token cookie automatically.
  return new EventSource(`/api/session/events`);
}

export function parseOpenCodeEvent(data: string): OpenCodeEvent | null {
  try {
    const raw = JSON.parse(data);
    if (raw && raw.payload && typeof raw.payload.type === "string") {
      return raw.payload as OpenCodeEvent;
    }
    if (raw && typeof raw.type === "string") {
      return raw as OpenCodeEvent;
    }
    return null;
  } catch {
    return null;
  }
}

// ── Multi-session dashboard ───────────────────────────

export interface SessionOverviewEntry {
  id: string;
  title: string;
  parentID: string;
  project_name: string;
  project_index: number;
  directory: string;
  is_busy: boolean;
  time: { created: number; updated: number };
  stats?: SessionStats;
}

export interface SessionsOverviewResponse {
  sessions: SessionOverviewEntry[];
  total: number;
  busy_count: number;
}

export interface SessionTreeNode {
  id: string;
  title: string;
  project_name: string;
  project_index: number;
  is_busy: boolean;
  stats?: SessionStats;
  children: SessionTreeNode[];
}

export interface SessionsTreeResponse {
  roots: SessionTreeNode[];
  total: number;
}

export async function fetchSessionsOverview(): Promise<SessionsOverviewResponse> {
  return apiFetch<SessionsOverviewResponse>("/sessions/overview");
}

export async function fetchSessionsTree(): Promise<SessionsTreeResponse> {
  return apiFetch<SessionsTreeResponse>("/sessions/tree");
}

// ── Context Window ────────────────────────────────────

export interface ContextItem {
  label: string;
  tokens: number;
}

export interface ContextCategory {
  name: string;
  label: string;
  tokens: number;
  pct: number;
  color: string;
  items: ContextItem[];
}

export interface ContextWindowResponse {
  context_limit: number;
  total_used: number;
  usage_pct: number;
  categories: ContextCategory[];
  estimated_messages_remaining: number | null;
}

export async function fetchContextWindow(
  sessionId?: string
): Promise<ContextWindowResponse | null> {
  try {
    const qs = sessionId ? `?session_id=${encodeURIComponent(sessionId)}` : "";
    return await apiFetch<ContextWindowResponse>(`/context-window${qs}`);
  } catch {
    return null;
  }
}

// ── File Edits (Diff Review) ──────────────────────────

export interface FileEditEntry {
  path: string;
  original_content: string;
  new_content: string;
  timestamp: string;
  index: number;
}

export interface FileEditsResponse {
  session_id: string;
  edits: FileEditEntry[];
  file_count: number;
}

export async function fetchFileEdits(sessionId: string): Promise<FileEditsResponse> {
  return apiFetch<FileEditsResponse>(
    `/session/${encodeURIComponent(sessionId)}/file-edits`
  );
}

// ── Cross-Session Search ──────────────────────────────

export interface SearchResultEntry {
  session_id: string;
  session_title: string;
  project_name: string;
  message_id: string;
  role: string;
  snippet: string;
  timestamp: number;
}

export interface SearchResponse {
  query: string;
  results: SearchResultEntry[];
  total: number;
}

export async function searchMessages(
  projectIdx: number,
  query: string,
  limit?: number
): Promise<SearchResponse> {
  const params = new URLSearchParams({ q: query });
  if (limit) params.set("limit", String(limit));
  return apiFetch<SearchResponse>(
    `/project/${projectIdx}/search?${params.toString()}`
  );
}

File: leptos-ui/src/api/session_views.rs
//! Session views API — matches React `api/session-views.ts`.

use crate::types::api::{
    SessionsOverviewResponse, SessionsTreeResponse, ContextWindowResponse,
    FileEditsResponse, SearchResponse,
};
use super::client::{api_fetch, ApiError};

// ── API functions ───────────────────────────────────────────────────

/// Fetch multi-session dashboard overview.
pub async fn fetch_sessions_overview() -> Result<SessionsOverviewResponse, ApiError> {
    api_fetch("/sessions/overview").await
}

/// Fetch session tree view.
pub async fn fetch_sessions_tree() -> Result<SessionsTreeResponse, ApiError> {
    api_fetch("/sessions/tree").await
}

/// Fetch context window usage for a session.
pub async fn fetch_context_window(session_id: Option<&str>) -> Result<ContextWindowResponse, ApiError> {
    let path = match session_id {
        Some(sid) => format!(
            "/context-window?session_id={}",
            js_sys::encode_uri_component(sid),
        ),
        None => "/context-window".to_string(),
    };
    api_fetch(&path).await
}

/// Fetch file edits for diff review.
pub async fn fetch_file_edits(session_id: &str) -> Result<FileEditsResponse, ApiError> {
    let path = format!(
        "/session/{}/file-edits",
        js_sys::encode_uri_component(session_id),
    );
    api_fetch(&path).await
}

/// Cross-session message search.
pub async fn search_messages(
    project_idx: usize,
    query: &str,
    limit: Option<usize>,
) -> Result<SearchResponse, ApiError> {
    let lim = limit.unwrap_or(50);
    let path = format!(
        "/project/{}/search?q={}&limit={}",
        project_idx,
        js_sys::encode_uri_component(query),
        lim,
    );
    api_fetch(&path).await
}
